spec/app_spec.rb

```diff
@@ -0,0 +1,116 @@
+require 'spec_helper'
+require 'rack/test'
+require 'omniauth-myusa'
+require 'stub_classes'
+
+describe 'ComplianceViewer' do
+  include Rack::Test::Methods
+
+  class StubComplianceData
+  end
+
+  def app
+    ComplianceViewer.new! StubComplianceData.new
+  end
+
+  describe '/' do
+    it 'returns as expected with the index if authed' do
+      StubComplianceData.any_instance.stubs(:keys).returns(%w(abc bcd))
+      get '/', {}, 'rack.session' => { user_email: 'example@example.com' }
+      expect(last_response).to be_ok
+      expect(last_response.body).to include('Projects')
+    end
+
+    it 'redirects to the auth page if unauthed' do
+      get '/'
+      expect(last_response.redirect?).to eq true
+      expect(last_response['Location']).to include('/auth/myusa')
+    end
+  end
+
+  describe '/results' do
+    it 'returns as expected with the index if authed' do
+      StubComplianceData.any_instance.stubs(:keys).returns(%w(abc bcd))
+      get '/results', {},
+          'rack.session' => { user_email: 'example@example.com' }
+      expect(last_response).to be_ok
+      expect(last_response.body).to include('Projects')
+    end
+
+    it 'redirects to the auth page if unauthed' do
+      get '/results'
+      expect(last_response.redirect?).to eq true
+      expect(last_response['Location']).to include('/auth/myusa')
+    end
+  end
+
+  describe '/results/:name' do
+    it 'returns successfully if passed a name that exists' do
+      name = 'good'
+      StubComplianceData.any_instance.stubs(:base_name).returns(name)
+      StubComplianceData.any_instance.stubs(:versions).returns(
+        [
+          StubClasses::StubObjectVersion.new(name),
+          StubClasses::StubObjectVersion.new(name),
+          StubClasses::StubObjectVersion.new(name)
+        ]
+      )
+      get "/results/#{name}", {},
+          'rack.session' => { user_email: 'example@example.com' }
+      expect(last_response).to be_ok
+      expect(last_response.body).to include(name)
+    end
+
+    it 'returns Invalid Project if passed a name that doesn\'t exist' do
+      StubComplianceData.any_instance.stubs(:versions).returns []
+      get "/results/name", {},
+          'rack.session' => { user_email: 'example@example.com' }
+      expect(last_response).to be_ok
+      expect(last_response.body).to include('Invalid Project')
+    end
+
+    it 'redirects to the auth page if unauthed with a name' do
+      get '/results/name'
+      expect(last_response.redirect?).to eq true
+      expect(last_response['Location']).to include('/auth/myusa')
+    end
+
+    it 'redirects to the auth page if unauthed without a name' do
+      get '/results/'
+      expect(last_response.redirect?).to eq true
+      expect(last_response['Location']).to include('/auth/myusa')
+    end
```

Comment:
This is intentional. The way routing works for Sinatra apps, `/results` and `/results/` are different, with `/results/` going to `/results/{name}`. That's why I put it in this block.